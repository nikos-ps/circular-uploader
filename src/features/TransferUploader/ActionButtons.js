import { memo } from "react";
import styled from "styled-components";

import Button from "components/Button";

const ActionsContainer = styled.div`
  display: flex;
  height: 100%;
  justify-content: center;
  align-items: center;
`;

const ActionButtons = memo(({ isLoaderActive, stopLoader, startLoader }) => {
  return (
    <ActionsContainer>
      {isLoaderActive ? (
        <Button secondary onClick={stopLoader}>
          End
        </Button>
      ) : (
        <Button onClick={startLoader}>Start</Button>
      )}
    </ActionsContainer>
  );
});

export default ActionButtons;
