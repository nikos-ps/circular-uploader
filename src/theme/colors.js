import { css } from "styled-components";

const colors = css`
  --colors-dodgerBlue: #409fff;
  --colors-scienceBlue: #0073e5;
  --colors-iron: #d4d7d9;
  --colors-woodSmoke: #17181a;
  --colors-silverSand: #babcbf;
  --colors-porcelain: #e8ebed;
  --colors-abbey: #484a4d;
  --colors-white: #ffffff;
`;

export default colors;
