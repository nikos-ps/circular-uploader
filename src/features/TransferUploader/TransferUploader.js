import { useEffect, useRef, useState, useCallback, useMemo } from "react";
import styled from "styled-components";

import Card from "components/Card";
import CircularLoader from "components/CircularLoader";
import Text from "components/Text";
import ActionButtons from "./ActionButtons";

const StatusText = styled(Text)`
  text-align: center;
  color: var(--colors-woodSmoke);
`;

const InfoText = styled(Text)`
  text-align: center;
  color: var(--colors-abbey);
  margin-top: 4px;
`;

const TransferUploader = () => {
  const [progress, setProgress] = useState(0);
  const [isTransfering, setIsTransfering] = useState(false);
  const intervalRef = useRef();

  const startLoader = useCallback(() => {
    setIsTransfering(true);

    intervalRef.current = setInterval(() => {
      setProgress((previous) => previous + 1);
    }, 500);
  }, []);

  const stopLoader = useCallback(() => {
    if (intervalRef.current) {
      clearInterval(intervalRef.current);
    }

    setProgress(0);
    setIsTransfering(false);
  }, []);

  useEffect(() => {
    if (progress === 100) {
      clearInterval(intervalRef.current);
    }
  }, [progress]);

  useEffect(() => {
    return () => clearInterval(intervalRef.current);
  }, []);

  const isLoaderActive = useMemo(() => progress !== 0, [progress]);

  return (
    <Card
      content={
        <>
          <CircularLoader radius={85} stroke={10} progress={progress} />
          {isTransfering && (
            <>
              <StatusText size="h2" as="h2">
                Transferring...
              </StatusText>
              <InfoText size="bodyCaption" as="p">
                ...Transfer related info...
              </InfoText>
            </>
          )}
        </>
      }
      actions={
        <ActionButtons
          isLoaderActive={isLoaderActive}
          stopLoader={stopLoader}
          startLoader={startLoader}
        />
      }
    />
  );
};

export default TransferUploader;
