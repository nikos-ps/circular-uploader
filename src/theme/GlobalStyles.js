import { createGlobalStyle } from "styled-components";
import colors from "./colors";

const GlobalStyles = createGlobalStyle`
  
  :root {
    ${colors}
  }
`;

export default GlobalStyles;
