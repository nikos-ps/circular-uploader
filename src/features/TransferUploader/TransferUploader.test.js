import { render, act } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { ThemeProvider } from "styled-components";

import { theme, GlobalStyles } from "theme";

import TransferUploader from "./TransferUploader";

describe("Features <TransferUploader />", () => {
  const renderUploader = () => {
    return render(
      <ThemeProvider theme={theme}>
        <GlobalStyles />
        <TransferUploader />
      </ThemeProvider>
    );
  };

  beforeEach(() => {
    jest.useFakeTimers();
  });

  afterEach(() => {
    jest.clearAllTimers();
    jest.useRealTimers();
  });

  it("renders correctly", () => {
    const { container } = renderUploader();

    expect(container).toMatchSnapshot();
  });

  it("has a button with text 'start' when loaded", () => {
    const { getByRole } = renderUploader();

    expect(getByRole("button")).toHaveTextContent("Start");
  });

  it("has a circular loader svg", () => {
    const { getByTestId } = renderUploader();

    expect(getByTestId("circular-loader")).toBeInTheDocument();
  });

  it("starts the loader when 'start' button is clicked", () => {
    const { getByRole } = renderUploader();

    userEvent.click(getByRole("button"));

    act(() => jest.advanceTimersByTime(500));

    expect(setInterval).toHaveBeenCalledTimes(1);
  });

  it("shows the 'end' button after the loader has started", () => {
    const { getByRole } = renderUploader();

    userEvent.click(getByRole("button"));

    act(() => jest.advanceTimersByTime(500));

    expect(getByRole("button")).toHaveTextContent("End");
  });

  it("updates the loader's progress label correctly", () => {
    const { getByRole, getByTestId } = renderUploader();

    userEvent.click(getByRole("button"));

    act(() => jest.advanceTimersByTime(1000));

    expect(getByTestId("progress-label")).toHaveTextContent(2);
  });

  it("resets the loader's progress label when 'end' button is clicked", () => {
    const { getByRole, getByTestId } = renderUploader();

    userEvent.click(getByRole("button"));

    act(() => jest.advanceTimersByTime(1000));

    userEvent.click(getByRole("button"));

    expect(getByTestId("progress-label")).toHaveTextContent(0);
  });
});
