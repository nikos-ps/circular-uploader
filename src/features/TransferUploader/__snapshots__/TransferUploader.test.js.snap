// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Features <TransferUploader /> renders correctly 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  background: var(--colors-white);
  border-radius: 10px;
  box-shadow: 0 0 12px 0 rgba(0,0,0,0.1),0 10px 30px 0 rgba(0,0,0,0.2);
  height: 410px;
  width: 280px;
}

.c4 {
  position: relative;
  margin-top: auto;
  border-top: 1px solid var(--colors-iron);
  height: 72px;
  padding: 0 20px;
}

.c1 {
  display: block;
  margin: 30px auto 12px;
  position: relative;
  width: 170px;
}

.c2 {
  border-radius: 10px;
  stroke-linecap: round;
  -webkit-transition: stroke-dashoffset 0.2s linear,stroke 0.2s linear;
  transition: stroke-dashoffset 0.2s linear,stroke 0.2s linear;
  -webkit-transform: rotate(-90deg);
  -ms-transform: rotate(-90deg);
  transform: rotate(-90deg);
  -webkit-transform-origin: 50% 50%;
  -ms-transform-origin: 50% 50%;
  transform-origin: 50% 50%;
  -webkit-animation: iwIftm 2s linear infinite;
  animation: iwIftm 2s linear infinite;
}

.c3 {
  color: var(--colors-woodSmoke);
  font-size: 3.875em;
  left: 50%;
  line-height: 1;
  margin: -2px 0 0 2px;
  position: absolute;
  top: 50%;
  -webkit-transform: translate(-50%,-50%);
  -ms-transform: translate(-50%,-50%);
  transform: translate(-50%,-50%);
  white-space: nowrap;
}

.c3 span {
  color: var(--colors-silverSand);
  display: inline-block;
  font-size: 22px;
  line-height: 1;
  position: relative;
  top: 9px;
  vertical-align: top;
}

.c6 {
  background: var(--colors-dodgerBlue);
  border: 1px solid var(--colors-dodgerBlue);
  border-radius: 1.25em;
  box-sizing: border-box;
  color: var(--colors-white);
  cursor: pointer;
  display: inline-block;
  font-size: 1em;
  height: 2.5em;
  font-weight: 500;
  -webkit-transition: border-color 0.2s cubic-bezier(0.77,0,0.175,1), background-color 0.2s cubic-bezier(0.77,0,0.175,1), color 0.2s cubic-bezier(0.77,0,0.175,1);
  transition: border-color 0.2s cubic-bezier(0.77,0,0.175,1), background-color 0.2s cubic-bezier(0.77,0,0.175,1), color 0.2s cubic-bezier(0.77,0,0.175,1);
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  width: 150px;
  outline: 0;
}

.c6:hover {
  background-color: var(--colors-scienceBlue);
  border-color: var(--colors-scienceBlue);
}

.c5 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  height: 100%;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

<div>
  <div
    class="c0"
  >
    <div>
      <div
        class="c1"
      >
        <svg
          data-testid="circular-loader"
          height="170"
          width="170"
        >
          <circle
            cx="85"
            cy="85"
            fill="transparent"
            r="80"
            stroke="var(--colors-porcelain)"
            stroke-dasharray="502.6548245743669 502.6548245743669"
            stroke-width="10"
          />
          <circle
            class="c2"
            cx="85"
            cy="85"
            fill="transparent"
            r="80"
            stroke="var(--colors-dodgerBlue)"
            stroke-dasharray="502.6548245743669 502.6548245743669"
            stroke-dashoffset="502.6548245743669"
            stroke-width="10"
          />
        </svg>
        <span
          class="c3"
          data-testid="progress-label"
        >
          0
          <span>
            %
          </span>
        </span>
      </div>
    </div>
    <div
      class="c4"
    >
      <div
        class="c5"
      >
        <button
          class="c6"
        >
          Start
        </button>
      </div>
    </div>
  </div>
</div>
`;
